src/backend/linux_raw/fs/raw_dir.rs

```diff
@@ -0,0 +1,204 @@
+///   [`RawDir::new`].
+pub struct RawDir<'buf, Fd: AsFd> {
+    fd: Fd,
+    buf: &'buf mut [MaybeUninit<u8>],
+    initialized: usize,
+    offset: usize,
+}
+
+impl<'buf, Fd: AsFd> RawDir<'buf, Fd> {
+    /// Create a new iterator from the given file descriptor and buffer.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # use std::mem::MaybeUninit;
+    /// # use rustix::fs::{cwd, Mode, OFlags, openat, RawDir};
+    ///
+    /// let fd = openat(cwd(), ".", OFlags::RDONLY | OFlags::DIRECTORY, Mode::empty()).unwrap();
+    ///
+    /// let mut buf = [MaybeUninit::uninit(); 2048];
+    /// for entry in RawDir::new(fd, &mut buf) {
+    ///     let entry = entry.unwrap();
+    ///     dbg!(&entry);
+    /// }
+    /// ```
+    ///
+    /// Contrived example that demonstrates reading entries with arbitrarily large file paths:
+    ///
+    /// ```
+    /// # use std::cmp::max;
+    /// # use std::mem::MaybeUninit;
+    /// # use rustix::fs::{cwd, Mode, OFlags, openat, RawDir};
+    /// # use rustix::io::Errno;
+    ///
+    /// let fd = openat(cwd(), ".", OFlags::RDONLY | OFlags::DIRECTORY, Mode::empty()).unwrap();
+    ///
+    /// // DO NOT DO THIS. Use `Vec::with_capacity(DIR_BUF_LEN)` to at least start the buffer
+    /// // off with *some* space.
+    /// let mut buf = Vec::new();
+    ///
+    /// 'read: loop {
+    ///     'resize: {
+    ///         for entry in RawDir::new(&fd, buf.spare_capacity_mut()) {
+    ///             let entry = match entry {
+    ///                 Err(Errno::INVAL) => break 'resize,
+    ///                 r => r.unwrap(),
+    ///             };
+    ///             dbg!(&entry);
+    ///         }
+    ///         break 'read;
+    ///     }
+    ///
+    ///     let new_capacity = max(buf.capacity() * 2, 1);
+    ///     buf.reserve(new_capacity);
+    /// }
+    /// ```
+    ///
+    /// Note that this is horribly inefficient as we'll most likely end up doing ~1 syscall per file.
+    pub fn new(fd: Fd, buf: &'buf mut [MaybeUninit<u8>]) -> Self {
+        Self {
+            fd,
+            buf,
+            initialized: 0,
+            offset: 0,
+        }
+    }
+}
+
+/// A raw directory entry, similar to `std::fs::DirEntry`.
+///
+/// Note that unlike the std version, this may represent the `.` or `..` entries.
+#[derive(Debug)]
+pub struct RawDirEntry<'a> {
+    file_name: &'a CStr,
+    file_type: u8,
+    inode_number: u64,
+    next_entry_cookie: i64,
+}
+
+impl<'a> RawDirEntry<'a> {
+    /// Returns the file name of this directory entry.
+    #[inline]
+    pub fn file_name(&self) -> &CStr {
+        self.file_name
+    }
+
+    /// Returns the type of this directory entry.
+    #[inline]
+    pub fn file_type(&self) -> FileType {
+        FileType::from_dirent_d_type(self.file_type)
+    }
+
+    /// Returns the inode number of this directory entry.
+    #[inline]
+    #[doc(alias = "ino")]
+    pub fn inode_number(&self) -> u64 {
```

Comment:
I like verbose names in general, however Rust's standard library calls a similar function [`ino`](https://doc.rust-lang.org/stable/std/os/unix/fs/trait.DirEntryExt.html#tymethod.ino), and the Unix field is named `d_ino`, so would it be ok to name this `ino()` for familiarity?

Reply:
Yup, sounds good.